pkg/scalers/splunk_observability_scaler.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Splunk metadata: %w", err)
+	}
+
+	apiClient, err := newSplunkO11yConnection(meta, logger)
+	if err != nil {
+		return nil, fmt.Errorf("error establishing Splunk Observability Cloud connection: %w", err)
+	}
+
+	return &splunkObservabilityScaler{
+		metadata:  meta,
+		apiClient: apiClient,
+		logger:    logger,
+	}, nil
+}
+
+func (s *splunkObservabilityScaler) getQueryResult() (float64, error) {
+	comp, err := s.apiClient.Execute(context.Background(), &signalflow.ExecuteRequest{
+		Program: s.metadata.Query,
+	})
+	if err != nil {
+		return -1, fmt.Errorf("error: could not execute signalflow query: %w", err)
+	}
+
+	s.logger.Info("Started MTS stream.")
+
+	time.Sleep(time.Duration(s.metadata.Duration * int(time.Second)))
+	if err := comp.Stop(context.Background()); err != nil {
+		return -1, fmt.Errorf("error creating SignalFlow client: %w", err)
+	}
+
+	s.logger.Info("Closed MTS stream.")
+
+	max := math.Inf(-1)
+	min := math.Inf(1)
+	valueSum := 0.0
+	valueCount := 0
+	s.logger.Info("Now iterating over results.")
+	for msg := range comp.Data() {
+		if len(msg.Payloads) == 0 {
+			s.logger.Info("No data retreived.")
+			continue
+		}
+		for _, pl := range msg.Payloads {
+			value, ok := pl.Value().(float64)
+			if !ok {
+				return -1, fmt.Errorf("error: could not convert Splunk Observability metric value to float64")
+			}
+			s.logger.Info(fmt.Sprintf("Encountering value %.4f\n", value))
+			max = math.Max(max, value)
+			min = math.Min(min, value)
+			valueSum += value
+			valueCount++
+		}
+	}
+
+	if valueCount > 1 && s.metadata.QueryAggregator == "" {
+		return 0, fmt.Errorf("query returned more than 1 series; modify the query to return only 1 series or add a queryAggregator")
+	}
+
+	switch s.metadata.QueryAggregator {
+	case "max":
+		s.logger.Info(fmt.Sprintf("Returning max value: %.4f\n", max))
+		return max, nil
+	case "min":
+		s.logger.Info(fmt.Sprintf("Returning min value: %.4f\n", min))
+		return min, nil
+	case "avg":
+		avg := valueSum / float64(valueCount)
+		s.logger.Info(fmt.Sprintf("Returning avg value: %.4f\n", avg))
+		return avg, nil
+	default:
+		return max, nil
+	}
+}
+
+func (s *splunkObservabilityScaler) GetMetricsAndActivity(_ context.Context, metricName string) ([]external_metrics.ExternalMetricValue, bool, error) {
+	num, err := s.getQueryResult()
+
+	if err != nil {
+		s.logger.Error(err, "error getting metrics from Splunk Observability Cloud.")
+		return []external_metrics.ExternalMetricValue{}, false, fmt.Errorf("error getting metrics from Splunk Observability Cloud: %w", err)
+	}
+	metric := GenerateMetricInMili(metricName, num)
+
+	return []external_metrics.ExternalMetricValue{metric}, num > s.metadata.ActivationTargetValue, nil
+}
+
+func (s *splunkObservabilityScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	metricName := kedautil.NormalizeString("signalfx")
+	re := regexp.MustCompile(`data\('([^']*)'`)
```

Comment:
Recompiling this regex every time `GetMetricSpecForScaling()` is expensive in terms of CPU operations. To optimize this. We should move the regex to be compiled with the package and not every time the function is called. To do this, we can simply put it in a var at the top of the package like so:

```go
var (
     dataRegex = regexp.MustCompile(`data\('([^']*)'`)
)
```

Is regex necessary though? Is there not a more predictable way to get the data returned?

Reply:
@circa10a this regex is used to extract the metric name from the SignalFlow query. How important is the metric name regarding KEDA operations? Could I just simply use smth like "signalfx-metric"?

Reply:
You just need a metric name per scaled object with a unique identifier. Take splunk's for example:

```go
// GetMetricSpecForScaling returns the MetricSpec for the Horizontal Pod Autoscaler
func (s *SplunkScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
	externalMetric := &v2.ExternalMetricSource{
		Metric: v2.MetricIdentifier{
			Name: GenerateMetricNameWithIndex(s.metadata.triggerIndex, kedautil.NormalizeString(fmt.Sprintf("splunk-%s", s.metadata.SavedSearchName))),
		},
		Target: GetMetricTarget(s.metricType, int64(s.metadata.TargetValue)),
	}
	metricSpec := v2.MetricSpec{
		External: externalMetric,
		Type:     externalMetricType,
	}
	return []v2.MetricSpec{metricSpec}
}
```

Above we ensure uniqueness using the trigger index, splunk-, and the saved search name. Here we could possibly do, trigger index, splunk-observability-, realm ?

More about the expectations of the [GetMetricSpecForScaling](https://github.com/kedacore/keda/blob/main/CREATE-NEW-SCALER.md#getmetricspecforscaling) function.

Reply:
Yeah, please change this as suggested.

Reply:
Removed regex. The metric name is now 'kedautil.NormalizeString("signalfx")'